libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1239 @@
+/* Copyright 2018 Endless Mobile, Inc.
+ *
+ * eos-discovery-feed is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * eos-discovery-feed is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with eos-discovery-feed.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include <errno.h>
+#include <math.h>
+#include <string.h>
+
+#include <gio/gio.h>
+
+#include <eos-shard/eos-shard-blob.h>
+#include <eos-shard/eos-shard-shard-file.h>
+#include <eos-shard/eos-shard-record.h>
+
+#include "feed-base-card-store.h"
+#include "feed-card-layout-direction.h"
+#include "feed-knowledge-app-artwork-card-store.h"
+#include "feed-knowledge-app-card-store.h"
+#include "feed-knowledge-app-news-card-store.h"
+#include "feed-knowledge-app-video-card-store.h"
+#include "feed-knowledge-app-proxy.h"
+#include "feed-orderable-model.h"
+#include "feed-quote-card-store.h"
+#include "feed-sizes.h"
+#include "feed-store-provider.h"
+#include "feed-text-sanitization.h"
+#include "feed-word-card-store.h"
+#include "feed-word-quote-card-store.h"
+
+typedef GPtrArray * (*ModelsFromResultsAndShardsFunc) (const char * const *shards,
+                                                       GPtrArray          *model_hts,
+                                                       gpointer            user_data);
+typedef GObject * (*ModelFromResultFunc) (GHashTable *model_ht,
+                                          gpointer    user_data);
+
+static GHashTable *
+hashtable_from_variant (GVariant *variant)
+{
+  GVariantIter iter;
+  gchar *key = NULL;
+  gchar *value = NULL;
+  g_autoptr(GHashTable) table = g_hash_table_new_full (g_str_hash,
+                                                       g_str_equal,
+                                                       g_free,
+                                                       g_free);
+
+  g_variant_iter_init (&iter, variant);
+
+  /* Since g_variant_iter_next allocates, we can pass the pointers
+   * directly to g_hash_table_insert without copying their contents */
+  while (g_variant_iter_next (&iter, "{ss}", &key, &value))
```

Comment:
At least according to valgrind and the documentation this is true. Perhaps it would be clearer to `g_strdup` and explicitly free as well, not sure.

Reply:
Yeah, that'd be OK